File: idea-live-templates/assets/examples/TitleCaseMacro.java
package com.example.livetemplates;

import com.intellij.codeInsight.template.Expression;
import com.intellij.codeInsight.template.ExpressionContext;
import com.intellij.codeInsight.template.Result;
import com.intellij.codeInsight.template.TemplateContextType;
import com.intellij.codeInsight.template.TextResult;
import com.intellij.codeInsight.template.macro.MacroBase;
import com.intellij.openapi.util.text.StringUtil;
import org.jetbrains.annotations.NotNull;

final class TitleCaseMacro extends MacroBase {

  TitleCaseMacro() {
    super("titleCase", "titleCase(String)");
  }

  private TitleCaseMacro(String name, String description) {
    super(name, description);
  }

  @Override
  protected Result calculateResult(Expression @NotNull [] params, ExpressionContext context, boolean quick) {
    String text = getTextResult(params, context, true);
    if (text == null) {
      return null;
    }
    if (!text.isEmpty()) {
      text = StringUtil.toTitleCase(text);
    }
    return new TextResult(text);
  }

  @Override
  public boolean isAcceptableInContext(TemplateContextType context) {
    return context instanceof MarkdownContext;
  }
}

File: idea-live-templates/assets/examples/TitleCaseMacro.kt
package com.example.livetemplates

import com.intellij.codeInsight.template.Expression
import com.intellij.codeInsight.template.ExpressionContext
import com.intellij.codeInsight.template.Result
import com.intellij.codeInsight.template.TemplateContextType
import com.intellij.codeInsight.template.TextResult
import com.intellij.codeInsight.template.macro.MacroBase
import com.intellij.openapi.util.text.StringUtil

internal class TitleCaseMacro : MacroBase("titleCase", "titleCase(String)") {

    override fun calculateResult(
        params: Array<out Expression>,
        context: ExpressionContext,
        quick: Boolean
    ): Result? {
        val text = getTextResult(params, context, true) ?: return null
        return TextResult(if (text.isNotEmpty()) StringUtil.toTitleCase(text) else text)
    }

    override fun isAcceptableInContext(context: TemplateContextType): Boolean {
        return context is MarkdownContext
    }
}
